astropy/modeling/core.py

```diff
@@ -3341,6 +3341,80 @@ def render(self, out=None, coords=None):
         return out
 
 
+    def replace_submodel(self, name=None, model=None):
+        """
+        Construct a new CompoundModel instance from an existing CompoundModel,
+        replacing the named submodel with the new model.
+
+        In order to ensure that inverses and names are kept/reconstructed, it's
+        necessary to rebuild the CompoundModel from the replaced node all the
+        way back to the base. The original CompoundModel is left untouched.
+
+        Parameters
+        ----------
+        model: CompoundModel instance
+            model to be reconstructed with replacement
+        name: str
+            name of submodel to be replaced
+        new_model: Model
+            replacement model
+        """
+        submodels = [m for m in self.traverse_postorder() if m.name == name]
+        if submodels:
+            if len(submodels) > 1:
+                raise ValueError(f"More than one submodel named {name}")
+
+            old_model = submodels.pop()
+            if len(old_model) != len(model):
+                raise ValueError("New and old models must have equal values "
+                                 "for n_inputs")
+
+            # Do this check first in order to raise a more helpful Exception,
+            # although it would fail trying to construct the new CompoundModel
+            if (old_model.n_inputs != model.n_inputs or
+                        old_model.n_outputs != model.n_outputs):
+                raise ValueError("New model must match numbers of inputs and "
+                                 "outputs of existing model")
+
+            tree = _get_submodel_tree(self, name)
+            while tree:
+                branch = self.copy()
```

Comment:
Maybe I misunderstand, but wouldn't it be more efficient to gather the needed information from the original model in one pass and rebuild the compound model then rather than keep drilling down the tree for each CompoundModel construction?

Reply:
I don't think I can use the original model because I may have replaced a `CompoundModel` with a `Model` and there may be a different number of elements (an entire set of branches may have been replaced by a single leaf). So I think I have to do the first iteration in order to stick my new model in its parent's `left` or `right` in order to get a new tree. Having done that once I don't need any new code to repeatedly do it, rather than go to the top and rebuild from the "directions" in `traverse_postorder()`, if that's what you're suggesting? Or is there a simpler trick I'm missing, I know you're the expert on the internal stucture of `CompoundModels`.